Add option to allow Select interaction logic to select overlapping features

bjornharrtell opened the pull request:
The original default behaviour of the Select interaction is to "Replace the currently selected feature(s) with the feature at the pixel". I made this code change to change the semantics to "Replace the currently selected feature(s) with the feature(s) at the pixel". I could not find a way to configure to get the desired behaviour with add/remove/toggle conditions).

This PR is a proposal with some unanswered questions. The "no change" condition was and is at fault since it only compares the first of the current selected features. I'm prepared to rework the PR if the questions can be resolved.

What is a sensible default behaviour? 

Should the default behaviour be configurable?

elemoine commented:
Anyone wants to review this for 3.2? I'll move it to v3.3.0 if not reviewed by tomorrow Wednesday.

fredj commented on src/ol/interaction/selectinteraction.js:
```diff
@@ -60,6 +60,13 @@ ol.interaction.Select = function(opt_options) {
+   * @private
+   * @type {boolean}
+   */
+  this.multi_ = goog.isDef(options.multi) ?
```
should be in a single line

fredj commented on externs/olx.js:
```diff
@@ -2414,6 +2414,15 @@ olx.interaction.SelectOptions.prototype.removeCondition;
+ * @type {boolean|undefined}
+ * @api
+ */
+olx.interaction.SelectOptions.prototype.multi;
```
the property must be added to the `olx.interaction.SelectOptions` typedef

bjornharrtell commented:
Thanks for the review @fredj, I've fixed those and squashed to a single commit. Travis CI build fails and I can get the same error locally but I don't understand what it means or how it can be related to my branch.

gberaudo commented on externs/olx.js:
```diff
@@ -2337,7 +2337,8 @@ olx.interaction.PinchZoomOptions.prototype.duration;
  *     removeCondition: (ol.events.ConditionType|undefined),
- *     toggleCondition: (ol.events.ConditionType|undefined)}}
+ *     toggleCondition: (ol.events.ConditionType|undefined),
+ *     multi: {boolean|undefined}}
```
Use parenthesis, not curly braces

bjornharrtell commented:
Thanks @gberaudo, don't know why I missed that. Fixed with updated squashed commit.

fredj commented on externs/olx.js:
```diff
@@ -2519,7 +2519,8 @@ olx.interaction.PointerOptions.prototype.handleUpEvent;
  *     removeCondition: (ol.events.ConditionType|undefined),
- *     toggleCondition: (ol.events.ConditionType|undefined)}}
+ *     toggleCondition: (ol.events.ConditionType|undefined),
+ *     multi: (boolean|undefined)}
```
a final `}` is missing

bjornharrtell commented:
Thanks @fredj, fixed missing curly brace, and that seem to have made Travis happy. Will try to avoid similar typos in the future.

fredj commented:
LVGTM, thanks @bjornharrtell

tschaub commented:
It looks like this introduced a regression in behavior.  See #3218 (and [a mailing list issue](https://groups.google.com/forum/#!topic/ol3-dev/9eOvA1_8K-I) which may or may not be related).

bjornharrtell commented:
The intention was to keep the existing behaviour and I when looking at the changes I don't understand what is causing the regression. The feature that gets selected should be the last one returned by forEachFeatureAtPixel both before and after this change.

fredj commented:
In the previous implementation, the first feature was returned by `map.forEachFeatureAtPixel` not the last (i just figured it out now).
The reason is that if something "truthy" is returned by the function the loop stops.

See #3221 for a fix
